src/Microsoft.AspNet.Mvc.Razor/Precompilation/RazorPreCompileModule.cs

```diff
@@ -0,0 +1,101 @@
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.Versioning;
+using Microsoft.AspNet.FileProviders;
+using Microsoft.Dnx.Compilation.CSharp;
+using Microsoft.Dnx.Runtime;
+using Microsoft.Framework.Caching.Memory;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc.Razor.Precompilation
+{
+    /// <summary>
+    /// An <see cref="ICompileModule"/> implementation that pre-compiles Razor views in the application.
+    /// </summary>
+    public abstract class RazorPreCompileModule : ICompileModule
+    {
+        private readonly IAssemblyLoadContext _loadContext;
+        private readonly object _memoryCacheLookupLock = new object();
+        private readonly Dictionary<PrecompilationCacheKey, MemoryCache> _memoryCacheLookup =
+            new Dictionary<PrecompilationCacheKey, MemoryCache>();
+
+        /// <summary>
+        /// Instantiates a new <see cref="RazorPreCompileModule"/> instance.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceProvider"/> for the application.</param>
+        public RazorPreCompileModule(IServiceProvider services)
+        {
+            _loadContext = services.GetRequiredService<IAssemblyLoadContext>();
+        }
+
+        /// <summary>
+        /// Gets or sets a value that determines if symbols (.pdb) file for the precompiled views is generated.
+        /// </summary>
+        public bool GenerateSymbols { get; protected set; }
+
+        /// <inheritdoc />
+        /// <remarks>Pre-compiles all Razor views in the application.</remarks>
+        public virtual void BeforeCompile(BeforeCompileContext context)
+        {
+            var fileProvider = new PhysicalFileProvider(context.ProjectContext.ProjectDirectory);
+
+            MemoryCache memoryCache;
+            lock (_memoryCacheLookupLock)
+            {
+                var cacheKey = new PrecompilationCacheKey
+                {
+                    Configuration = context.ProjectContext.Configuration,
+                    TargetFramework = context.ProjectContext.TargetFramework
+                };
+
+                if (!_memoryCacheLookup.TryGetValue(cacheKey, out memoryCache))
+                {
+                    // When CompactOnMemoryPressure is true, the MemoryCache evicts items at every gen2 collection.
+                    // In DTH, gen2 happens frequently enough to make it undesirable for caching precompilation results. We'll
+                    // disable listening for memory pressure for the MemoryCache instance used by precompilation.
+                    memoryCache = new MemoryCache(new MemoryCacheOptions { CompactOnMemoryPressure = false });
+
+                    _memoryCacheLookup[cacheKey] = memoryCache;
+                }
+            }
+
+            var viewCompiler = new RazorPreCompiler(
+                context,
+                _loadContext,
+                fileProvider,
+                memoryCache)
```

Comment:
MemoryCache is limited to using string keys - so we're stuck with either creating a composite key using the tfm and configuration - or varying the instance in itself. This seemed cleaner.